Add audit information for job & pipeline pauses

rjinskidepop opened the pull request:
Signed-off-by: Richard Inskip <richard.inskip@depop.com>

<!--
Thanks for submitting a pull request to Concourse!

If you haven't already, feel free to [add yourself] as a contributor so that
you can add labels to your PR and re-trigger its builds if they fail.

Also check the [PR requirements] if you haven't already!
-->

[add yourself]: https://github.com/concourse/governance#individual-contributors
[PR requirements]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md#pull-request-requirements

## Changes proposed by this PR



<!--
Summarize your changes as a checklist, leaving any unfinished work as unchecked
items. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->

* [x] move `jobs.pause` and `pipelines.pause` columns to their own database tables
* [x] add `paused_by` and `paused_at` when a job is paused
* [x] add `paused_by` and `paused_at` when a pipeline is paused
* [x] expose `paused_by` and `paused_at` in API (`GET jobs`, `GET job`,`GET pipeline`, `GET pipelines`)
* [x] add `paused_by` and `paused_at` visual in the UI
* [x] add fly commands `paused-jobs` and `paused-pipelines` to list paused jobs & pipelines with new meta data 

## Notes to reviewer

![image](https://user-images.githubusercontent.com/77615995/125457852-0912abf2-0520-4038-b018-924fce035e1d.png)

![image](https://user-images.githubusercontent.com/77615995/125457889-51f0e005-b2b6-4c62-8341-d717209c8d2e.png)

![image](https://user-images.githubusercontent.com/77615995/125457904-f472ae97-db0c-4bac-82d9-246078eb2128.png)

- Elm/front end is most definitely not my strong point - this is the first time I've read or written it - and couldn't really see where/how to add any tests. I may need some guidance if that is a requirement for getting merged.

- I spoke with @aoldershaw on discord about the increase to the SQL join limit - this was due to the increased complexity in `buildsQuery` in `atc/db/build.go`

- I went back and forth on adding this meta data in a separate resource/API/object rather than including on a `Job` or `Pipeline` but found this solution much cleaner and less disruptive to the client side code. 

## Release Note

* Add pipeline and job pause meta information - who and when.

Commit:
Add audit information for job & pipeline pauses

Signed-off-by: Richard Inskip <richard.inskip@depop.com>

EstebanFS commented on atc/postgresrunner/join_limit_conn.go:
```diff
@@ -12,16 +12,17 @@ import (
 func validateJoinLimit(query string) error {
-	// Each subquery can have up to 8 explicit JOINs
+	// Each subquery can have up to <limit> explicit JOINs
+	limit := 10
 	individualQueries := ExtractQueries(query)
 	for _, q := range individualQueries {
 		numJoins := strings.Count(strings.ToUpper(q), "JOIN")
-		if numJoins > 8 {
+		if numJoins > limit {
 			errMsg := "the following query has too many JOINs\n\n" + q
 			if len(individualQueries) > 1 {
 				errMsg += "\n\nit is part of this full query:\n\n" + query
 			}
-			errMsg += fmt.Sprintf("\n\njoin_collapse_limit defaults to 8 JOINs, while this query has %d. exceeding the limit can result in really slow queries", numJoins)
+			errMsg += fmt.Sprintf("\n\njoin_collapse_limit defaults to %d JOINs, while this query has %d. exceeding the limit can result in really slow queries", limit, numJoins)
 
 			return fmt.Errorf(errMsg)
 		}
```
We shouldn't modify this test, the idea is to make sure we are always using the default configuration on Postgres (Postgres has a default limit of 8).

EstebanFS commented on atc/db/migration/migrations/1626194317_add_pipeline_pauses_table.up.sql:
```diff
@@ -0,0 +1,21 @@
+WHERE
+  pipelines.paused = TRUE;
+
+ALTER TABLE pipelines
+  DROP COLUMN paused;
```
We would want to keep this column around, so we don't need to JOIN with the new table everytime (specially when we don't need the extra information that table provides), that way we can avoid the JOIN limit of 8 that postgres has.

EstebanFS commented on atc/db/migration/migrations/1626194317_add_pipeline_pauses_table.up.sql:
```diff
@@ -0,0 +1,21 @@
+CREATE TABLE pipeline_pauses (
+  pipeline_id integer PRIMARY KEY,
+  paused boolean NOT NULL DEFAULT TRUE,
```
since we are going to keep the column in the original table, we won't need this attribute anymore

EstebanFS reviewed (changes requested):
Thanks for all the effort you are putting on making this PR possible.

Since the suggestions I made would impact all the code, I would leave the comments like this for now, and then when we implement the changes I could take another look.

Please don't hesitate on ping me if you have any questions about it.

rjinskidepop commented:
👋🏻  @EstebanFS 

Thank you for taking the time to review. 

- RE: the joins. I was trying to avoid having two sources of truth `job.paused` and `job_pauses.paused` to avoid situation where they may become misaligned. Maybe this can be avoided with the use of a trigger. Let me investigate and come up with some ideas.

- The `JobPauseRequest` was forward planning to support other features. I like the ability when pinning resources that the user is able to comment and I had ideas of a TTL or a reminder. Appreciate they may never see the light of day, but, could it be useful to keep this construct in?

Let me know your thoughts and I'll start tweaking the uncoupled suggestions 👍🏻

EstebanFS commented:
Hey @rjinskidepop thanks for all your hard work on this, about the two points you made, I talked with @clarafu about it:

- I don't like having two source of truth either, but we think that could be the easiest and safest solution to solve the problem and continue with the development, also since those tables will be probably pretty small we are no too worried about that (we just need to find a good way to keep the consistency)

- About the Structure thing, We understand that could be the beginnig of some other features in the future, but since we are not sure how Concourse would grow we prefer to keep it as simple and consistent as the rest of the project. If by any chance we need to implement some new feature that affect that portion of code, we can always refactor to get some better code.

I'm looking forward to see how this feature ends.

rjinskidepop commented:
I have been doing some thinking about the table structure. One of the advantages of this approach was that adding `Job.PausedBy` and `Job.PauseAt` on the `Job` object (and similarly for `Pipeline`) mean't the client side changes were much simpler and coherent. I did experiment with a new API/resource but, didn't like it. It seemed too involved for the value add.

I think to keep things simple *and* avoid the join the right approach would be to move `paused_by` and `paused_at` onto the `jobs` and `pipelines` tables and remove the _new_table entirely. If/when more complex requirements come along, this separation can be reinstantiated and move the metadata off `Job` and `Pipeline` and onto its own resource/object/API and create a clear separation.

I'm going to work on that assumption and start the changes; let me know if you have any objections

rjinskidepop commented on fly/commands/paused_jobs.go:
```diff
@@ -0,0 +1,105 @@
+		return nil
+	}
+
+	pausedJobs := make([]atc.Job, 0)
+	for _, j := range jobs {
+		if j.Paused {
+			pausedJobs = append(pausedJobs, j)
+		}
+	}
```
I can indeed. The structure of `paused_pipelines` was consistent with `pipelines` and `paused_jobs` with `jobs` but will align `paused_*`

Commit:
PR feedback

Signed-off-by: Richard Inskip <richard.inskip@depop.com>

EstebanFS commented:
Great Job @rjinskidepop, thanks for taking some time to fix all the comments, the new code looks great so far

EstebanFS reviewed (changes requested):
@clarafu and I just added a few easy suggestions here, thanks for all your hard work on this

rjinskidepop commented on fly/commands/paused_pipelines.go:
```diff
@@ -0,0 +1,115 @@
+	pausedPipelines := make([]atc.Pipeline, 0)
+	for _, p := range pipelines {
+		if p.Paused {
+			pipelines = append(pipelines, p)
```
Good spot 👍🏻

rjinskidepop commented on atc/db/job.go:
```diff
@@ -104,8 +106,28 @@ type Job interface {
+	"j.disable_manual_trigger",
+	"j.paused_by",
+	"j.paused_at").
+	From("jobs j").
+	LeftJoin("pipelines p ON j.pipeline_id = p.id").
```
Nice, I wasn't sure how to achieve this - I'm not familiar with squirrel. I was originally trying `.From("jobs j").From("pipelines p")` which sent me down a rabbit hole 🤓

Commit:
PR feedback

Signed-off-by: Richard Inskip <richard.inskip@depop.com>

rjinskidepop commented:
> @clarafu and I just added a few easy suggestions here, thanks for all your hard work on this

Thank you @EstebanFS ; all valid suggestions. Reflected and updated 👍🏻

EstebanFS reviewed (approved):
LGTM!
